scripts/openwebnet-discovery.nse

```diff
@@ -0,0 +1,246 @@
+  [12] = "F453AV",
+  [13] = "H4684",
+  [15] = "F427 (Gateway Open-KNX)",
+  [16] = "F453",
+  [23] = "H4684",
+  [27] = "L4686SDK",
+  [44] = "MH200N",
+  [51] = "F454",
+  [200] = "F454 (new?)"
+}
+
+local who = {
+  [0] = "Scenarios",
+  [1] = "Lighting",
+  [2] = "Automation",
+  [3] = "Power Management",
+  [4] = "Heating",
+  [5] = "Burglar Alarm",
+  [6] = "Door Entry System",
+  [7] = "Multimedia",
+  [9] = "Auxiliary",
+  [13] = "Device Communication",
+  [14] = "Light+shutters actuators lock",
+  [15] = "CEN",
+  [16] = "Sound System",
+  [17] = "Scenario Programming",
+  [18] = "Energy Management",
+  [24] = "Lighting Management",
+  [25] = "CEN plus",
+  [1000] = "Diagnostic",
+  [1001] = "Automation Diagnostic",
+  [1004] = "Heating Diagnostic",
+  [1008] = "Door Entry System Diagnostic",
+  [1013] = "Device Diagnostic"
+}
+
+local device_dimension = {
+  ["Time"] = "0",
+  ["Date"] = "1",
+  ["IP Address"] = "10",
+  ["Net Mask"] = "11",
+  ["MAC Address"] = "12",
+  ["Device Type"] = "15",
+  ["Firmware Version"] = "16",
+  ["Hardware Version"] = "17",
+  ["Uptime"] = "19",
+  ["Micro Version"] = "20",
+  ["Date and Time"] = "22",
+  ["Kernel Version"] = "23",
+  ["Distribution Version"] = "24",
+  ["Gateway IP address"] = "50",
+  ["DNS IP address 1"] = "51",
+  ["DNS IP address 2"] = "52"
+}
+
+local ACK = "*#*1##"
+local NACK = "*#*0##"
+
+-- Initiates a socket connection
+-- Returns the socket and error message
+local function get_socket(host, port, request)
+
+  local sd, response, early_resp = comm.opencon(host, port, request)
+
+  if sd == nil then
+    return nil, "Socket connection error."
+  end
+
+  if not response then
+    return nil, "Poor internet connection or no response."
+  end
+
+  if response == NACK then
+    return nil, "Received a negative ACK as response."
+  end
+
+  return sd, nil
+end
+
+local function get_response(sd, request)
+
+  local res = {}
+  local status, data
+
+  sd:send(request)
+
+  repeat
+    status, data = sd:receive_buf("##", true)
+    if status and data ~= ACK then
+      table.insert(res, data)
+    end
+    if data == ACK then
+      break
+    end
+
+    -- If response is NACK, it means the request method is not supported
+    if data == NACK then
+      res = {}
+    end
+  until not status
+
+  return res
+end
+
+local function format_dimensions(res)
+
+  if res["Date and Time"] then
+    res["Date"] = string.match(res["Date and Time"], "((%d+)%.(%d+)%.(%d+))$")
+
+    res["Time"] = string.match(res["Date and Time"], "^((%d+)%.(%d+)%.(%d+))")
+    res["Time"] = string.gsub(res["Time"], "%.", ":")
+
+    res["Date and Time"] = nil
+  end
+
+  if res["Device Type"] then
+    res["Device Type"] = device[ tonumber( res["Device Type"] ) ]
+  end
+
+  if res["MAC Address"] then
+    local t = {}
+    for _, v in ipairs(stdnse.strsplit("%.%s*", res["MAC Address"])) do
+      table.insert(t, stdnse.tohex(tonumber(v)))
+    end
+
+    res["MAC Address"] = table.concat(t, ":")
```

Comment:
Instead of splitting, converting, and joining, try using `string.gsub` with a function like so:

```lua
formatted = string.gsub(raw, "(%d+)%.?",
    function (number)
      return do_something(number)
    end
)
```

You have options for `do_something`: you could simply return the byte with the value of `number`, which would convert the whole string to a 6-byte blob suitable for passing to `stdnse.format_mac`. You could convert it directly to hex with `string.format`, and change the separator to ":" either there or in a separate pass (not consuming it in the pattern). Etc.